File: ios/SurfriderDataCards/Tally/Components.swift
//
//  The controls the eight screens are built from.
//
//  Everything here is the design's, and the pressed states are the part worth
//  keeping: the handoff is explicit that every control tints from the accent
//  ramp and that no browser -- or in this case system -- default is used. A
//  SwiftUI Button's default is a fade to 50% opacity, which is not the same
//  thing and reads as the control going away rather than being held.
//

import SwiftUI
import UIKit

// MARK: - Feedback

/// The part of a tap you feel before you can see anything.
///
/// Screen 6 is a keypad somebody works at for an hour, and a keypad with no
/// haptic reads as lag even when the digit lands in the same frame -- the thing
/// it is being compared against is the system keyboard, which has one. This is
/// the cheapest speed there is: nothing gets faster, but the tap stops feeling
/// like it went unanswered.
///
/// Prepared rather than fired cold. The Taptic Engine takes a beat to wake, and
/// the press that would land in it is the first of the session -- the one that
/// sets what the screen feels like.
enum Haptics {
    private static let key = UIImpactFeedbackGenerator(style: .light)
    private static let step = UIImpactFeedbackGenerator(style: .medium)

    static func prepare() {
        key.prepare()
        step.prepare()
    }

    /// A digit. Softer than the system keyboard's, because there are hundreds.
    static func tap() {
        key.impactOccurred(intensity: 0.7)
        key.prepare()
    }

    /// A value committed and the screen moving on.
    static func advance() {
        step.impactOccurred()
        step.prepare()
    }
}

// MARK: - Buttons

/// Outlined in the accent. The affirmative action on every screen.
///
/// On iOS 26 the outline stays and the inside becomes accent-tinted glass; the
/// label lightens to `accent300` because accent-on-accent-tint is the one place
/// the ramp does not carry. See Glass.swift.
struct PrimaryButtonStyle: ButtonStyle {
    var minHeight: CGFloat = 52
    var size: CGFloat = 16
    @Environment(\.isEnabled) private var isEnabled

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(Nocturne.Face.label(size, weight: .medium))
            .foregroundStyle(Nocturne.hasGlass ? Nocturne.accent300 : Nocturne.accent)
            // Room inside the outline. Most of these stretch and never notice,
            // but the one that does not -- "Start fresh", beside a button that
            // does -- was drawing its label flush against both edges.
            .padding(.horizontal, 16)
            .frame(maxWidth: .infinity, minHeight: minHeight)
            .controlSurface(
                RoundedRectangle(
                    cornerRadius: Nocturne.hasGlass ? Nocturne.Radius.glass : Nocturne.Radius.base,
                    style: .continuous
                ),
                tint: Nocturne.glassTint,
                fill: Nocturne.accent.opacity(configuration.isPressed ? 0.22 : 0),
                stroke: Nocturne.hasGlass ? Nocturne.accent.opacity(0.55) : Nocturne.accent
            )
            .opacity(isEnabled ? 1 : 0.45)
    }
}

/// Outlined in the divider. The way out, or the lesser of two actions.
struct SecondaryButtonStyle: ButtonStyle {
    var minHeight: CGFloat = 50
    var size: CGFloat = 15
    @Environment(\.isEnabled) private var isEnabled

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(Nocturne.Face.label(size, weight: .medium))
            .foregroundStyle(Nocturne.text)
            .padding(.horizontal, 16)
            .frame(maxWidth: .infinity, minHeight: minHeight)
            .controlSurface(
                RoundedRectangle(
                    cornerRadius: Nocturne.hasGlass ? Nocturne.Radius.glass : Nocturne.Radius.base,
                    style: .continuous
                ),
                fill: Nocturne.text.opacity(configuration.isPressed ? 0.14 : 0),
                stroke: Nocturne.divider
            )
            .opacity(isEnabled ? 1 : 0.45)
    }
}

/// Accent text with no border: the nav bar, and the small inline actions.
struct TextButtonStyle: ButtonStyle {
    var size: CGFloat = 16
    @Environment(\.isEnabled) private var isEnabled

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .font(Nocturne.Face.label(size))
            .foregroundStyle(Nocturne.accent)
            .opacity(configuration.isPressed ? 0.6 : (isEnabled ? 1 : 0.45))
    }
}

// MARK: - Chrome

/// The back affordance: a leading caret and the destination word, in accent.
///
/// Not a title bar. Screen titles are large and flush-left in the content, and
/// this sits above them in a 44pt bar of its own.
struct NavBar<Trailing: View>: View {
    let backLabel: String?
    let onBack: (() -> Void)?
    @ViewBuilder var trailing: Trailing

    init(back: String? = nil, onBack: (() -> Void)? = nil, @ViewBuilder trailing: () -> Trailing = { EmptyView() }) {
        self.backLabel = back
        self.onBack = onBack
        self.trailing = trailing()
    }

    var body: some View {
        HStack(spacing: 0) {
            if let backLabel, let onBack {
                Button(action: onBack) {
                    HStack(spacing: 2) {
                        Image(systemName: Nocturne.Icon.back).font(.system(size: 15, weight: .medium))
                        Text(backLabel)
                    }
                }
                .buttonStyle(ChromeButtonStyle())
            }
            Spacer(minLength: 0)
            trailing
        }
        .frame(height: Nocturne.navBar)
        .padding(.horizontal, 14)
    }
}

/// Uppercase, tracked, accent. Sits above a title.
struct Kicker: View {
    let text: String
    var icon: String?
    var color: Color = Nocturne.accent

    var body: some View {
        HStack(spacing: 6) {
            if let icon { Image(systemName: icon).font(.system(size: 11)) }
            Text(text.uppercased())
                .font(Nocturne.Face.label(11))
                .tracking(1.2)
        }
        .foregroundStyle(color)
    }
}

/// A section label in the body: uppercase, tracked, text at 45%.
struct SectionLabel: View {
    let text: String

    var body: some View {
        Text(text.uppercased())
            .font(Nocturne.Face.label(11))
            .tracking(1.1)
            .foregroundStyle(Nocturne.text(45))
    }
}

/// The progress bar. 6pt while reading, 4pt while checking.
struct ProgressLine: View {
    let fraction: Double
    var height: CGFloat = 6

    var body: some View {
        GeometryReader { geo in
            ZStack(alignment: .leading) {
                Capsule().fill(Nocturne.track)
                Capsule()
                    .fill(
                        LinearGradient(
                            colors: [Nocturne.accent700, Nocturne.accent, Nocturne.accent300],
                            startPoint: .leading, endPoint: .trailing
                        )
                    )
                    // The travelled part carries a little of its own light, so
                    // the bar is legible against the ambient ground rather than
                    // competing with it.
                    .shadow(color: Nocturne.accent.opacity(Nocturne.hasGlass ? 0.6 : 0), radius: 7)
                    .frame(width: max(0, min(1, fraction)) * geo.size.width)
            }
        }
        .frame(height: height)
        .animation(.easeOut(duration: 0.35), value: fraction)
    }
}

/// A freestanding rule, fading to transparent over its outer 48pt. A Nocturne
/// signature: box outlines and in-control separators stay solid, these do not.
struct FadingRule: View {
    var body: some View {
        GeometryReader { geo in
            let fade = min(48, geo.size.width / 3)
            LinearGradient(
                stops: [
                    .init(color: .clear, location: 0),
                    .init(color: Nocturne.divider, location: fade / geo.size.width),
                    .init(color: Nocturne.divider, location: 1 - fade / geo.size.width),
                    .init(color: .clear, location: 1),
                ],
                startPoint: .leading, endPoint: .trailing
            )
        }
        .frame(height: 1)
    }
}

/// A form field: 12pt label, optional suffix at 55%, and a 46pt input.
struct Field<Content: View>: View {
    let label: String
    var optional: Bool = false
    @ViewBuilder var content: Content

    private var fieldRadius: CGFloat {
        Nocturne.hasGlass ? Nocturne.Radius.glass : Nocturne.Radius.base
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 5) {
            HStack(spacing: 5) {
                Text(label).foregroundStyle(Nocturne.text(70))
                if optional { Text("optional").foregroundStyle(Nocturne.text(55)) }
            }
            .font(Nocturne.Face.label(12))

            content
                .font(Nocturne.Face.label(16))
                .foregroundStyle(Nocturne.text)
                .tint(Nocturne.accent)
                .padding(.horizontal, 12)
                .frame(minHeight: 46)
                // NOT GLASS, on either path. A field is something to read and
                // type into, and a lens over the ground is the wrong answer to
                // "where does the text sit" -- it is the flat surface the
                // design specifies. Only the corner follows the glass world, so
                // a form and the button under it are the same family.
                .background(
                    RoundedRectangle(cornerRadius: fieldRadius, style: .continuous)
                        .fill(Nocturne.surface)
                )
                .overlay(
                    RoundedRectangle(cornerRadius: fieldRadius, style: .continuous)
                        .stroke(Nocturne.divider, lineWidth: 1)
                )
        }
    }
}

/// The accent-tinted panel: draft card, tally note, anything the accent frames
/// rather than floods.
struct TintedPanel<Content: View>: View {
    var radius: CGFloat = Nocturne.Radius.base
    @ViewBuilder var content: Content

    var body: some View {
        content
            .background(RoundedRectangle(cornerRadius: shape, style: .continuous).fill(Nocturne.accent900))
            .overlay(RoundedRectangle(cornerRadius: shape, style: .continuous).stroke(Nocturne.accent700, lineWidth: 1))
    }

    /// NOT GLASS. This panel is the one thing on screen that holds buttons of
    /// its own, and glass inside glass is mud -- the draft card came out a flat
    /// lavender block with two paler blocks sitting in it. It is also the place
    /// the design is most explicit that the accent FRAMES rather than floods.
    /// So the fill stays flat and the buttons in it keep the depth.
    private var shape: CGFloat {
        Nocturne.hasGlass ? max(radius, Nocturne.Radius.glass) : radius
    }
}

/// A row in a list, with the design's 1pt bottom divider and press tint.
struct RowButton<Content: View>: View {
    let action: () -> Void
    @ViewBuilder var content: Content

    var body: some View {
        Button(action: action) {
            content
                .frame(maxWidth: .infinity, alignment: .leading)
                .contentShape(Rectangle())
        }
        .buttonStyle(RowButtonStyle())
        .overlay(alignment: .bottom) {
            Rectangle().fill(Nocturne.divider).frame(height: 1)
        }
    }
}

struct RowButtonStyle: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .background(Nocturne.text.opacity(configuration.isPressed ? 0.05 : 0))
    }
}

// MARK: - Layout

extension View {
    /// The page margin, everywhere.
    func pageMargin() -> some View { padding(.horizontal, Nocturne.margin) }

    /// Pin an action over a fade to the ground, as screen 1 does.
    func bottomFade() -> some View {
        background(
            LinearGradient(
                stops: [
                    .init(color: .clear, location: 0),
                    .init(color: Nocturne.ground, location: 0.28),
                    .init(color: Nocturne.ground, location: 1),
                ],
                startPoint: .top, endPoint: .bottom
            )
        )
    }
}

/// Every screen's ground and safe padding, in one place.
struct ScreenBody<Content: View>: View {
    var ground: Color = Nocturne.ground
    var topPadding: CGFloat = Nocturne.safeTop
    @ViewBuilder var content: Content

    var body: some View {
        VStack(spacing: 0) { content }
            .frame(maxWidth: .infinity, maxHeight: .infinity, alignment: .top)
            .padding(.top, topPadding)
            .nocturneGround(ground)
    }
}

// MARK: - Tags

/// The small claim a control makes about itself: which reader filled a box in,
/// or that a path is unproven.
struct Tag: View {
    let text: String
    var size: CGFloat = 11
    var tracking: CGFloat = 0

    var body: some View {
        Text(text)
            .font(Nocturne.Face.label(size, weight: .medium))
            .tracking(tracking)
            .foregroundStyle(Nocturne.text)
            .padding(.horizontal, 8)
            .padding(.vertical, 3)
            // Flat on both paths: a tag is always inside something -- a line of
            // body text, or a button's label -- and never on the bare ground,
            // so there is nothing behind it for a lens to find.
            .background(Capsule().fill(Nocturne.accent800))
    }
}

File: android/app/src/main/java/com/mateobesse/surfriderdatacards/tally/Components.kt
//
//  The controls the eight screens are built from.
//
//  Components.swift, in Compose. The pressed states are the part worth keeping:
//  the handoff is explicit that every control tints from the accent ramp and
//  that no platform default is used. Android's default is the ripple, which is
//  a different gesture from "being held", so every control here draws its own
//  pressed state and passes `indication = null`.
//

package com.mateobesse.surfriderdatacards.tally

import android.os.Build
import android.view.HapticFeedbackConstants
import android.view.View
import androidx.compose.animation.core.FastOutSlowInEasing
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.animation.core.tween
import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.BoxScope
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.ColumnScope
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.RowScope
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.navigationBarsPadding
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.statusBarsPadding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.text.BasicTextField
import androidx.compose.foundation.text.KeyboardOptions
import androidx.compose.material3.Icon
import androidx.compose.material3.LocalContentColor
import androidx.compose.material3.LocalTextStyle
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableIntStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.alpha
import androidx.compose.ui.draw.clip
import androidx.compose.ui.draw.drawBehind
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Brush
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.SolidColor
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.layout.onSizeChanged
import androidx.compose.ui.platform.LocalDensity
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp

// Feedback

/**
 * The part of a tap you feel before you can see anything.
 *
 * Screen 6 is a keypad somebody works at for an hour, and a keypad with no
 * haptic reads as lag even when the digit lands in the same frame. These are
 * the system's own keyboard constants, so they follow the phone's touch
 * feedback setting rather than overriding it.
 */
object Haptics {
    /** A digit. */
    fun tap(view: View) {
        view.performHapticFeedback(HapticFeedbackConstants.KEYBOARD_TAP)
    }

    /** A value committed and the screen moving on. */
    fun advance(view: View) {
        view.performHapticFeedback(
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.R) HapticFeedbackConstants.CONFIRM
            else HapticFeedbackConstants.VIRTUAL_KEY,
        )
    }
}

// Buttons

private enum class ButtonKind { Primary, Secondary }

/**
 * Outlined in the accent. The affirmative action on every screen.
 *
 * Stretches, like the SwiftUI style it comes from; pass `stretch = false` for
 * the one place a button is the width of its label.
 */
@Composable
fun PrimaryButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    minHeight: Dp = 52.dp,
    size: Int = 16,
    stretch: Boolean = true,
    content: @Composable RowScope.() -> Unit,
) = NocturneButton(ButtonKind.Primary, onClick, modifier, enabled, minHeight, size, stretch, content)

/** Outlined in the divider. The way out, or the lesser of two actions. */
@Composable
fun SecondaryButton(
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    minHeight: Dp = 50.dp,
    size: Int = 15,
    stretch: Boolean = true,
    content: @Composable RowScope.() -> Unit,
) = NocturneButton(ButtonKind.Secondary, onClick, modifier, enabled, minHeight, size, stretch, content)

@Composable
private fun NocturneButton(
    kind: ButtonKind,
    onClick: () -> Unit,
    modifier: Modifier,
    enabled: Boolean,
    minHeight: Dp,
    size: Int,
    stretch: Boolean,
    content: @Composable RowScope.() -> Unit,
) {
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    val shape = RoundedCornerShape(Nocturne.Radius.base)

    val fill = when (kind) {
        ButtonKind.Primary -> Nocturne.accent.copy(alpha = if (pressed) 0.22f else 0f)
        ButtonKind.Secondary -> Nocturne.text.copy(alpha = if (pressed) 0.14f else 0f)
    }
    val stroke = if (kind == ButtonKind.Primary) Nocturne.accent else Nocturne.divider
    val label = if (kind == ButtonKind.Primary) Nocturne.accent else Nocturne.text

    Row(
        modifier = modifier
            .then(if (stretch) Modifier.fillMaxWidth() else Modifier)
            .defaultMinSize(minHeight = minHeight)
            .alpha(if (enabled) 1f else 0.45f)
            .clip(shape)
            .background(fill)
            .border(1.dp, stroke, shape)
            .clickable(interaction, indication = null, enabled = enabled, role = Role.Button, onClick = onClick)
            // Room inside the outline -- see the note in Components.swift.
            .padding(horizontal = 16.dp, vertical = 8.dp),
        horizontalArrangement = Arrangement.spacedBy(9.dp, Alignment.CenterHorizontally),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        CompositionLocalProvider(
            LocalContentColor provides label,
            LocalTextStyle provides Nocturne.Face.label(size, FontWeight.Medium),
        ) { content() }
    }
}

/** An icon sized to sit beside a button's label. */
@Composable
fun ButtonIcon(icon: ImageVector) {
    Icon(icon, contentDescription = null, modifier = Modifier.size(20.dp))
}

/** Accent text with no border: the small inline actions. */
@Composable
fun TextAction(text: String, onClick: () -> Unit, size: Int = 16, enabled: Boolean = true) {
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    Text(
        text,
        style = Nocturne.Face.label(size),
        color = Nocturne.accent,
        modifier = Modifier
            .alpha(if (pressed) 0.6f else if (enabled) 1f else 0.45f)
            .clickable(interaction, indication = null, enabled = enabled, role = Role.Button, onClick = onClick)
            .padding(vertical = 6.dp),
    )
}

// Chrome

/**
 * Accent text in the nav bar. Padded out to the minimum touch target, because
 * on the nav bar it is the only thing there is to hit.
 */
@Composable
fun ChromeButton(onClick: () -> Unit, size: Int = 16, enabled: Boolean = true, content: @Composable RowScope.() -> Unit) {
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    Row(
        modifier = Modifier
            .defaultMinSize(minHeight = Nocturne.minTap)
            .alpha(if (pressed) 0.6f else if (enabled) 1f else 0.45f)
            .clickable(interaction, indication = null, enabled = enabled, role = Role.Button, onClick = onClick)
            .padding(horizontal = 8.dp),
        horizontalArrangement = Arrangement.spacedBy(2.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        CompositionLocalProvider(
            LocalContentColor provides Nocturne.accent,
            LocalTextStyle provides Nocturne.Face.label(size),
        ) { content() }
    }
}

/**
 * The back affordance: a leading caret and the destination word, in accent.
 *
 * Not a title bar. Screen titles are large and flush-left in the content, and
 * this sits above them in a bar of its own.
 */
@Composable
fun NavBar(back: String? = null, onBack: (() -> Unit)? = null, trailing: @Composable RowScope.() -> Unit = {}) {
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .height(Nocturne.navBar)
            .padding(horizontal = 8.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        if (back != null && onBack != null) {
            ChromeButton(onBack) {
                Icon(Nocturne.Icon.back, contentDescription = null, modifier = Modifier.size(24.dp))
                Text(back)
            }
        }
        Spacer(Modifier.weight(1f))
        trailing()
    }
}

/** Uppercase, tracked, accent. Sits above a title. */
@Composable
fun Kicker(text: String, icon: ImageVector? = null, color: Color = Nocturne.accent) {
    Row(horizontalArrangement = Arrangement.spacedBy(6.dp), verticalAlignment = Alignment.CenterVertically) {
        if (icon != null) Icon(icon, contentDescription = null, tint = color, modifier = Modifier.size(13.dp))
        Text(text.uppercase(), style = Nocturne.Face.label(11, tracking = 1.2.sp), color = color)
    }
}

/** A section label in the body: uppercase, tracked, text at 45%. */
@Composable
fun SectionLabel(text: String, modifier: Modifier = Modifier) {
    Text(text.uppercase(), style = Nocturne.Face.label(11, tracking = 1.1.sp), color = Nocturne.text(45), modifier = modifier)
}

/** The progress bar. 6dp while reading, 4dp while checking. */
@Composable
fun ProgressLine(fraction: Float, modifier: Modifier = Modifier, height: Dp = 6.dp) {
    val shown by animateFloatAsState(
        targetValue = fraction.coerceIn(0f, 1f),
        animationSpec = tween(350, easing = FastOutSlowInEasing),
        label = "progress",
    )
    val fill = remember { Brush.horizontalGradient(listOf(Nocturne.accent700, Nocturne.accent, Nocturne.accent300)) }
    Canvas(modifier.fillMaxWidth().height(height)) {
        val radius = CornerRadius(size.height / 2, size.height / 2)
        drawRoundRect(Nocturne.track, cornerRadius = radius)
        if (shown > 0f) {
            drawRoundRect(fill, size = Size(size.width * shown, size.height), cornerRadius = radius)
        }
    }
}

/** A form field: 12sp label, optional suffix at 55%, and a 46dp input. */
@Composable
fun Field(label: String, modifier: Modifier = Modifier, optional: Boolean = false, content: @Composable BoxScope.() -> Unit) {
    val shape = RoundedCornerShape(Nocturne.Radius.base)
    Column(modifier, verticalArrangement = Arrangement.spacedBy(5.dp)) {
        Row(horizontalArrangement = Arrangement.spacedBy(5.dp)) {
            Text(label, style = Nocturne.Face.label(12), color = Nocturne.text(70))
            if (optional) Text("optional", style = Nocturne.Face.label(12), color = Nocturne.text(55))
        }
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .defaultMinSize(minHeight = 46.dp)
                .clip(shape)
                .background(Nocturne.surface)
                .border(1.dp, Nocturne.divider, shape)
                .padding(horizontal = 12.dp),
            contentAlignment = Alignment.CenterStart,
        ) {
            CompositionLocalProvider(
                LocalContentColor provides Nocturne.text,
                LocalTextStyle provides Nocturne.Face.label(16),
            ) { content() }
        }
    }
}

/**
 * A text input inside a `Field`. Flat, on the surface the design specifies;
 * the placeholder is text at 35%, which is what a system field would draw.
 */
@Composable
fun FieldInput(
    value: String,
    onValueChange: (String) -> Unit,
    placeholder: String,
    keyboardOptions: KeyboardOptions = KeyboardOptions.Default,
) {
    BasicTextField(
        value = value,
        onValueChange = onValueChange,
        singleLine = true,
        textStyle = Nocturne.Face.label(16).copy(color = Nocturne.text),
        cursorBrush = SolidColor(Nocturne.accent),
        keyboardOptions = keyboardOptions,
        modifier = Modifier.fillMaxWidth(),
        decorationBox = { inner ->
            Box(contentAlignment = Alignment.CenterStart) {
                if (value.isEmpty()) Text(placeholder, style = Nocturne.Face.label(16), color = Nocturne.text(35))
                inner()
            }
        },
    )
}

/** The accent-tinted panel: draft card, tally note, anything the accent frames rather than floods. */
@Composable
fun TintedPanel(modifier: Modifier = Modifier, radius: Dp = Nocturne.Radius.base, content: @Composable ColumnScope.() -> Unit) {
    val shape = RoundedCornerShape(radius)
    Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(shape)
            .background(Nocturne.accent900)
            .border(1.dp, Nocturne.accent700, shape),
        content = content,
    )
}

/** A card the size of its content: the flat outlined panel. */
@Composable
fun Panel(modifier: Modifier = Modifier, content: @Composable ColumnScope.() -> Unit) {
    val shape = RoundedCornerShape(Nocturne.Radius.base)
    Column(
        modifier = modifier
            .fillMaxWidth()
            .clip(shape)
            .border(1.dp, Nocturne.divider, shape),
        content = content,
    )
}

/** A row in a list, with the design's 1dp bottom divider and press tint. */
@Composable
fun RowButton(onClick: () -> Unit, modifier: Modifier = Modifier, content: @Composable RowScope.() -> Unit) {
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    Row(
        modifier = modifier
            .fillMaxWidth()
            .background(Nocturne.text.copy(alpha = if (pressed) 0.05f else 0f))
            .clickable(interaction, indication = null, role = Role.Button, onClick = onClick)
            .drawBehind {
                drawRect(Nocturne.divider, topLeft = Offset(0f, size.height - 1.dp.toPx()), size = Size(size.width, 1.dp.toPx()))
            },
        verticalAlignment = Alignment.CenterVertically,
        content = content,
    )
}

/**
 * The small claim a control makes about itself: which reader filled a box in,
 * or that a path is unproven.
 */
@Composable
fun Tag(text: String, size: Int = 11, tracking: Float = 0f) {
    Text(
        text,
        style = Nocturne.Face.label(size, FontWeight.Medium, tracking = tracking.sp),
        color = Nocturne.text,
        modifier = Modifier
            .clip(RoundedCornerShape(percent = 50))
            .background(Nocturne.accent800)
            .padding(horizontal = 8.dp, vertical = 3.dp),
    )
}

// Layout

/** The page margin, everywhere. */
fun Modifier.pageMargin(): Modifier = padding(horizontal = Nocturne.margin)

/** Every screen's ground and its clearance under the status bar, in one place. */
@Composable
fun ScreenBody(
    ground: Color = Nocturne.ground,
    topPadding: Dp = Nocturne.belowStatusBar,
    content: @Composable ColumnScope.() -> Unit,
) {
    Column(
        modifier = Modifier
            .fillMaxSize()
            .background(ground)
            .statusBarsPadding()
            .padding(top = topPadding),
        content = content,
    )
}

/**
 * A scrolling body with its actions floated over the foot of it.
 *
 * The SwiftUI screens attach these with `.safeAreaInset(edge: .bottom)`, so
 * that a long list scrolls past the button rather than stopping short of it.
 * This is the same arrangement: the actions are measured, and the content is
 * handed their height to leave clear at the end of its scroll.
 */
@Composable
fun ColumnScope.WithPinnedActions(
    actions: @Composable ColumnScope.() -> Unit,
    ground: Color = Nocturne.ground,
    content: @Composable (bottomClearance: Dp) -> Unit,
) {
    var actionsHeight by remember { mutableIntStateOf(0) }
    val clearance = with(LocalDensity.current) { actionsHeight.toDp() }

    Box(Modifier.weight(1f).fillMaxWidth()) {
        content(clearance)

        Column(
            modifier = Modifier
                .align(Alignment.BottomCenter)
                .fillMaxWidth()
                .onSizeChanged { actionsHeight = it.height }
                // `linear-gradient(to top, ground 72%, transparent)`: a list
                // dissolves before it reaches the button rather than being cut
                // at its edge.
                .background(Brush.verticalGradient(0f to Color.Transparent, 0.28f to ground, 1f to ground))
                .navigationBarsPadding()
                .pageMargin()
                .padding(top = 16.dp, bottom = Nocturne.aboveNavBar),
            verticalArrangement = Arrangement.spacedBy(10.dp),
            horizontalAlignment = Alignment.CenterHorizontally,
            content = actions,
        )
    }
}

/** A column that fills the space above the pinned actions and scrolls. */
@Composable
fun BottomClearance(clearance: Dp) {
    Spacer(Modifier.height(clearance + 12.dp).width(1.dp))
}
